samples/matter/common/src/matter_init.cpp

```diff
@@ -0,0 +1,230 @@
+
+#if defined(CONFIG_NET_L2_OPENTHREAD)
+	sInitResult = ThreadStackMgr().InitThreadStack();
+	VerifyInitResultOrReturn(sInitResult, "ThreadStackMgr().InitThreadStack() failed");
+
+	sInitResult = ConfigureThreadRole();
+	VerifyInitResultOrReturn(sInitResult, "Cannot configure Thread role");
+
+#elif defined(CONFIG_CHIP_WIFI)
+	if (!sCommissioningInstance) {
+		sInitResult = CHIP_ERROR_INTERNAL;
+		VerifyInitResultOrReturn(sInitResult, "No valid commissioning instance");
+	}
+	sCommissioningInstance->Init();
+#else
+	sInitResult = CHIP_ERROR_INTERNAL;
+	VerifyInitResultOrReturn(sInitResult, "No valid L2 network backend selected");
+#endif /* CONFIG_NET_L2_OPENTHREAD */
+
+#ifdef CONFIG_CHIP_OTA_REQUESTOR
+	/* OTA image confirmation must be done before the factory data init. */
+	OtaConfirmNewImage();
+#endif
+
+#ifdef CONFIG_MCUMGR_TRANSPORT_BT
+	/* Initialize DFU over SMP */
+	GetDFUOverSMP().Init();
+	GetDFUOverSMP().ConfirmNewImage();
+#endif
+
+	/* Initialize CHIP server */
+#if CONFIG_CHIP_FACTORY_DATA
+	VerifyOrReturn(sFactoryDataProvider, LOG_ERR("No valid factory data"));
+	sInitResult = sFactoryDataProvider->Init();
+	VerifyInitResultOrReturn(sInitResult, "FactoryDataProvider::Init() failed");
+
+	SetDeviceInstanceInfoProvider(sFactoryDataProvider);
+	SetDeviceAttestationCredentialsProvider(sFactoryDataProvider);
+	SetCommissionableDataProvider(sFactoryDataProvider);
+#else
+	SetDeviceInstanceInfoProvider(&DeviceInstanceInfoProviderMgrImpl());
+	SetDeviceAttestationCredentialsProvider(Examples::GetExampleDACProvider());
+#endif
+
+	VerifyOrReturn(sServerInitParams, LOG_ERR("No valid server initialization parameters"));
+	sInitResult = sServerInitParams->InitializeStaticResourcesBeforeServerInit();
+	VerifyInitResultOrReturn(sInitResult, "InitializeStaticResourcesBeforeServerInit() failed");
+
+	sInitResult = Server::GetInstance().Init(*sServerInitParams);
+	VerifyInitResultOrReturn(sInitResult, "Server::Init() failed");
+
+	ConfigurationMgr().LogDeviceConfig();
+	PrintOnboardingCodes(RendezvousInformationFlags(chip::RendezvousInformationFlag::kBLE));
+	AppFabricTableDelegate::Init();
+
+	sInitResult = PlatformMgr().AddEventHandler(reinterpret_cast<PlatformManager::EventHandlerFunct>(arg), 0);
+	VerifyInitResultOrReturn(sInitResult, "Cannot register CHIP event handler");
+
+	if (sCustomInit) {
+		sInitResult = sCustomInit();
+		VerifyInitResultOrReturn(sInitResult, "Custom initialization failed");
+	}
+}
+} // namespace
+
+/* Public API */
+namespace Nordic
+{
+namespace Matter
+{
+	CHIP_ERROR InitChipServer(PlatformManager::EventHandlerFunct chipEventHandler, Optional<ChipInitData> initData)
+	{
+		if (initData.HasValue()) {
+			auto customInitData = initData.Value();
+			sCommissioningInstance = customInitData.mNetworkingInstance;
+			sCustomInit = customInitData.mCustomInit;
+			sServerInitParams = customInitData.mServerInitParams;
+#if CONFIG_CHIP_FACTORY_DATA
+			sFactoryDataProvider = customInitData.mFactoryDataProvider;
+#endif
```

Comment:
Do we need custom init data at all? We didn't have it at all previously. I guess, customers don't need a custom implementation because they can just ignore calling the DoInitChipServer method in their apptask.cpp code and implement it on its own. Moreover, it makes the code more complicated because a customer needs to understand the whole "custom" dependencies first, and it makes our solution less attractive. If they saw only one method "InitChipServer" (located anywhere) they would have understood it without thinking about any other stuff.

Reply:
The thing is we had, but this was hidden in each sample initialization sequence. On example might be deferred attribute persister in the light bulb or customized factory data provider in the weather station app. Regarding customers perspective, I guess we need to find a consensus. One approach is to hard-code everything with #ifdefs just to support our current samples structure. The second one is to make everything configurable, so may as well do nothing with the current state. I guess none of these is what we want. The singular `InitChipServer` won't work because it gives no means to inject customized CHIP interfaces implementation. Also we need to keep in mind that we have at least 2 threads here that need to be synchronized. Otherwise our solution is prone to races, which actually happened in the past.

Reply:
Discussed f2f.